[CodeCompletion] Enable ASTScope in code completion

rintaro opened the pull request:
Enable ASTScope lookup (by default) in code completion.
rdar://problem/58939376

For fast-completion where the body of `AbstractFunctionDecl` might be replaced with another body parsed from another source buffer:
* Use `getOriginalBodySourceRange()` for `AbstractFunctionBodyScope`
* Remember the replaced range in `SourceManager`
* Translate source ranges for finding scopes when the lookup location is in the replaced range (`findInnermostEnclosingScope()`)
* Bypass source range checks when the child scope from the replaced range (`widenSourceRangeForChildren()` and `verifyThatChildrenAreContainedWithin `)
* Clear `ASTScope` for each completion

Unrelated to fast completion:
* Add generic params and where clause scope even with missing body
* Be lenient with `ASTScope` / `DeclContext` mismatch in code completion mode

rintaro committed:
[CodeCompletion Remember the replaced source range in SourceManager

rintaro committed:
[CodeCompletion] Enable ASTScope in code completion

* Re-create `ASTScope` for each completion
* Add generic params and where clause scope even without missing body
* Use `getOriginalBodySourceRange()` for `AbstractFunctionBodyScope`
* Source range translations for replaced ranges when finding scopes
* Bypass source range checks when the completion happens in the replaced
  range
* Be lenient with ASTScope / DeclContext mismatch in code completion

rintaro commented:
@swift-ci Please smoke test

rintaro commented:
@swift-ci Please smoke test

slavapestov commented on lib/AST/ASTScopeLookup.cpp:
```diff
@@ -82,6 +82,14 @@ const ASTScopeImpl *ASTScopeImpl::findStartingScopeForLookup(
   if (!startingScope) {
+
+    // Be lenient in code completion mode. There are cases where the decl
+    // context doesn't match with the ASTScope. e.g. dangling attributes.
```
Can you add a FIXME and file a radar for me or Dave? The ASTScope should be a 'single source of truth', even for invalid code.

rintaro replied:
Filed rdar://problem/66945121

slavapestov commented on lib/AST/ASTScopeSourceRange.cpp:
```diff
@@ -67,6 +67,14 @@ ASTScopeImpl::widenSourceRangeForChildren(const SourceRange range,
   auto r = range;
+
+  // HACK: For code completion. If the range of the child is from another
+  // source buffer, don't widen using that range.
```
Can we try to clean this up as well?

slavapestov commented on lib/AST/ASTScopeCreation.cpp:
```diff
@@ -1609,6 +1604,12 @@ ASTScopeImpl *GenericTypeOrExtensionWholePortion::expandScope(
+
+  // Prevent circular request bugs caused by illegal input and
+  // doing lookups that getExtendedNominal in the midst of getExtendedNominal.
+  if (scope->shouldHaveABody() && !scope->doesDeclHaveABody())
```
Can this happen in non-code completion cases too?

rintaro replied:
Actually, yes. For example, for:
```swift
extension Collection where Element == Int
```
current master emits 
```
test.swift:2:42: error: expected '{' in extension
extension Collection where Element == Int
                                         ^
test.swift:2:28: error: cannot find type 'Element' in scope
extension Collection where Element == Int
                           ^~~~~~~
test.swift:2:36: error: generic signature requires types '<<error type>>' and 'Int' to be the same
extension Collection where Element == Int
                                   ^
```
With this change, it's just:
```
test.swift:2:42: error: expected '{' in extension
extension Collection where Element == Int
                                         ^
```

rintaro replied:
Which file can I add a test case for this?

rintaro replied:
Added to `test/decl/ext/extensions.swift`

slavapestov commented:
This is great, even with the hacks, which I think we can try to sort out later. The only remaining usage of non-ASTScope lookup is now in lldb.

rintaro committed:
[Test] Add a test case for extension where clause with missing body

rdar://problem/66943328

rintaro commented:
@swift-ci Please smoke test

davidungar commented on lib/AST/ASTScopeLookup.cpp:
```diff
@@ -152,24 +175,24 @@ ASTScopeImpl::findChildContaining(SourceLoc loc,
-      return 0 >= ASTScopeImpl::compare(loc, scope->getSourceRangeOfScope(),
-                                        sourceMgr,
-                                        /*ensureDisjoint=*/false);
+      return !(*this)(scope, loc);
```
Could you please double-check this? (Or maybe just reassure me) What about the == case?

rintaro replied:
Thanks for notifying! I think your concern is legitimate (the comparison function should not return `true` for `==` case).
But, it turned out that this function is not called. According to https://en.cppreference.com/w/cpp/algorithm/lower_bound , comparison happens only by `comp(element, value)` , but not `comp(value, element)`. 

> The signature of the predicate function should be equivalent to the following:
> ```
> bool pred(const Type1 &a, const Type2 &b);
> ```
> The type `Type1` must be such that an object of type `ForwardIt` can be dereferenced and then implicitly converted to `Type1`. The type `Type2` must be such that an object of type `T` can be implicitly converted to `Type2`

So I wrote this block to:
```cpp
  auto *const *child = llvm::lower_bound(
      getChildren(), loc,
      [&sourceMgr](const ASTScopeImpl *scope, SourceLoc loc) {
         ...
      });
```

davidungar replied:
Thank you for addressing it. I worry that a future maintainer, (or even myself) will get confused. If the function on line 183 is not called, do we still need to supply it? If so, can we just put an llvm unreachable in it? Otherwise, how about a comment around line 184 explaining away the potential confusion?

rintaro replied:
Hmm, since I removed `operator()(SourceLoc loc, const ASTScopeImpl *scope)` and rewrote `CompareLocs` with a lambda expression, I believe there's no unreachable code anymore.

davidungar reviewed (commented):
Overall, looks pretty good. Just one comment where I could use some reassurance, because the code seems to assume that !(a<b) == b<a . Thanks for doing this!

rintaro committed:
[ASTScope] Simplify binary search code in findChildContaining()

The comparision happens only by 'comp(element, value)' not
'comp(value, element)'. We only need
`operator()(const ASTScopeImpl *, SourceLoc)`.

Use llvm::lower_bound() instead of std::lower_bound().

rintaro commented:
@swift-ci Please smoke test

rintaro commented:
@swift-ci Please smoke test

davidungar commented on lib/AST/ASTScopeLookup.cpp:
```diff
@@ -143,33 +151,40 @@ bool ASTScopeImpl::checkSourceRangeOfThisASTNode() const {
-  if (child != getChildren().end() &&
-      sourceMgr.rangeContainsTokenLoc((*child)->getSourceRangeOfScope(), loc))
-    return *child;
+  auto *const *child = llvm::lower_bound(
+      getChildren(), loc,
+      [&sourceMgr](const ASTScopeImpl *scope, SourceLoc loc) {
+        ASTScopeAssert(scope->checkSourceRangeOfThisASTNode(), "Bad range.");
+        auto rangeOfScope = scope->getSourceRangeOfScope();
+        loc = translateLocForReplacedRange(sourceMgr, rangeOfScope, loc);
+        return -1 == ASTScopeImpl::compare(rangeOfScope, loc, sourceMgr,
+                                           /*ensureDisjoint=*/false);
+      });
+
+  if (child != getChildren().end()) {
+    auto rangeOfScope = (*child)->getSourceRangeOfScope();
+    loc = translateLocForReplacedRange(sourceMgr, rangeOfScope, loc);
+    if (sourceMgr.rangeContainsTokenLoc(rangeOfScope, loc))
+      return *child;
+  }
```
Great! Thanks.

davidungar reviewed (approved):
Great! Thanks.
